presto-spi/src/main/java/com/facebook/presto/spi/block/DictionaryBlock.java

```diff
@@ -464,4 +465,50 @@ public DictionaryBlock compact()
             return this;
         }
     }
+
+    @Override
+    public BlockLease flatten(ArrayAllocator allocator)
+    {
+        Block block = dictionary;
+        int[] map = ids;
+        int[] leasedMap = null;
+        while (true) {
+            if (block instanceof DictionaryBlock) {
+                DictionaryBlock dictionaryBlock = (DictionaryBlock) block;
+                int[] ids = dictionaryBlock.ids;
+
+                if (leasedMap != null) {
+                    allocator.returnArray(leasedMap);
+                }
+                int[] newMap = allocator.getIntArray(positionCount);
+                leasedMap = newMap;
+                for (int i = 0; i < positionCount; ++i) {
+                    newMap[i] = ids[map[i] + dictionaryBlock.idsOffset];
+                }
+                map = newMap;
+
+                block = dictionaryBlock.dictionary;
+            }
+            else if (block instanceof RunLengthEncodedBlock) {
+                RunLengthEncodedBlock rle = (RunLengthEncodedBlock) block;
+                int[] newMap = allocator.getIntArray(positionCount);
+                Arrays.fill(newMap, 0, 0, positionCount);
```

Comment:
This line doesn't seem to do anything. Why are you filling from index 0 to 0?

Reply:
Great catch, this was a bug.  It would have manifested if there was prior filled data from the allocator that was being used, which wasn't in the tests (all of them request fresh data).  I added a test that explicitly filled in prior data, and also made it so that the tests share the allocator (to increase the chances of catching this sort of thing).  The real intent is to transform an RLE:

N = count, V = value

RLE(N x V)

Into:

Dictionary([0,0,...N], [V])